File: src/org/elixir_lang/code_insight/lookup/element_renderer/CallDefinitionClause.java
package org.elixir_lang.code_insight.lookup.element_renderer;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.codeInsight.lookup.LookupElementPresentation;
import com.intellij.navigation.ItemPresentation;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.PsiElement;
import com.intellij.ui.RowIcon;
import org.elixir_lang.Icons;
import org.elixir_lang.call.Visibility;
import org.elixir_lang.psi.call.Call;
import org.jetbrains.annotations.NotNull;

public class CallDefinitionClause extends com.intellij.codeInsight.lookup.LookupElementRenderer<LookupElement> {
    /*
     * Fields
     */

    @NotNull
    private final String name;

    /*
     * Constructors
     */

    public CallDefinitionClause(@NotNull String name) {
        this.name = name;
    }

    /*
     * Public Instance Methods
     */

    @Override
    public void renderElement(LookupElement element, LookupElementPresentation presentation) {
        presentation.setItemText(name);
        presentation.setItemTextBold(true);

        PsiElement psiElement = element.getPsiElement();

        if (psiElement == null) {
            renderObject(element);
        } else {
            renderPsiElement(psiElement, presentation);
        }
    }

    /*
     * Private Instance Methods
     */

    private void renderCall(@NotNull Call call, @NotNull LookupElementPresentation presentation) {
        if (org.elixir_lang.psi.CallDefinitionClause.is(call)) {
            renderCallDefinitionClause(call, presentation);
        }
    }

    private void renderCallDefinitionClause(@NotNull Call call, @NotNull LookupElementPresentation presentation) {
        org.elixir_lang.structure_view.element.CallDefinitionClause structureView =
                org.elixir_lang.structure_view.element.CallDefinitionClause.Companion.fromCall(call);

        if (structureView != null) {
            renderStructureView(structureView, presentation);
        }
    }

    private void renderItemPresentation(@NotNull ItemPresentation itemPresentation,
                                        @NotNull LookupElementPresentation lookupElementPresentation) {
        lookupElementPresentation.setIcon(itemPresentation.getIcon(true));
        String presentableText = itemPresentation.getPresentableText();

        if (presentableText != null) {
            int nameLength = name.length();
            int presentableTextLength = presentableText.length();

            if (nameLength <= presentableTextLength) {
                lookupElementPresentation.appendTailText(presentableText.substring(nameLength), true);
            }
        }

        String locationString = itemPresentation.getLocationString();

        if (locationString != null) {
            lookupElementPresentation.appendTailText(" (" + locationString + ")", false);
        }
    }

    private void renderObject(@NotNull LookupElement lookupElement) {
        Logger logger = Logger.getInstance(CallDefinitionClause.class);

        Object object = lookupElement.getObject();
        String title  = "CallDefinitionClause render called on LookupElement with null getPsiElement";
        String message = "## name\n" +
                "\n" +
                "```\n" +
                name + "\n" +
                "```\n" +
                "\n" +
                "## getObject()\n"+
                "\n" +
                "### toString()\n" +
                "\n" +
                "```\n" +
                object + "\n" +
                "```\n" +
                "\n" +
                "### getClass().getName()\n" +
                "\n" +
                "```\n" +
                object.getClass().getName() + "\n" +
                "```\n";

        logger.error(message, new Throwable(title));
    }

    private void renderPsiElement(@NotNull PsiElement psiElement, @NotNull LookupElementPresentation presentation) {
        if (psiElement instanceof Call) {
            renderCall((Call) psiElement, presentation);
        } else if (psiElement instanceof org.elixir_lang.beam.psi.impl.CallDefinitionImpl) {
            renderCallDefinitionImpl(
                    (org.elixir_lang.beam.psi.impl.CallDefinitionImpl<?>) psiElement,
                    presentation
            );
        }
    }

    private void renderCallDefinitionImpl(
            @NotNull org.elixir_lang.beam.psi.impl.CallDefinitionImpl<?> callDefinition,
            @NotNull LookupElementPresentation presentation) {
        int arity = callDefinition.getNameArityInterval().getArityInterval().getMinimum();
        presentation.appendTailText("/" + arity, true);

        // Mirror the source render path's icon (navigation.item_presentation.CallDefinitionHead):
        // a RowIcon of [time (function/macro), visibility, call-definition-clause].  Only exported
        // (public) definitions are offered in completion, so visibility is PUBLIC.
        RowIcon icon = new RowIcon(3);
        icon.setIcon(Icons.Time.from(callDefinition.getTime()), 0);
        icon.setIcon(Icons.Visibility.from(Visibility.PUBLIC), 1);
        icon.setIcon(Icons.CALL_DEFINITION_CLAUSE, 2);
        presentation.setIcon(icon);
    }

    private void renderStructureView(@NotNull org.elixir_lang.structure_view.element.CallDefinitionClause structureView,
                                     @NotNull LookupElementPresentation presentation) {
        renderItemPresentation(structureView.getPresentation(), presentation);
    }
}

File: src/org/elixir_lang/code_insight/lookup/element_renderer/CallDefinitionClause.kt
package org.elixir_lang.code_insight.lookup.element_renderer

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.codeInsight.lookup.LookupElementPresentation
import com.intellij.codeInsight.lookup.LookupElementRenderer
import com.intellij.navigation.ItemPresentation
import com.intellij.openapi.diagnostic.Logger
import com.intellij.psi.PsiElement
import org.elixir_lang.semantic.semantic

class CallDefinitionClause(private val name: String) : LookupElementRenderer<LookupElement>() {
    override fun renderElement(element: LookupElement, presentation: LookupElementPresentation) {
        presentation.itemText = name
        presentation.isItemTextBold = true
        val psiElement = element.psiElement
        if (psiElement == null) {
            renderObject(element)
        } else {
            renderElement(psiElement, presentation)
        }
    }

    private fun renderElement(psiElement: PsiElement, presentation: LookupElementPresentation) {
        when (val semantic = psiElement.semantic) {
            is org.elixir_lang.semantic.call.definition.Clause -> renderElement(semantic.presentation, presentation)
            else -> Unit
        }
    }

    private fun renderElement(
        itemPresentation: ItemPresentation,
        lookupElementPresentation: LookupElementPresentation
    ) {
        lookupElementPresentation.icon = itemPresentation.getIcon(true)
        val presentableText = itemPresentation.presentableText
        if (presentableText != null) {
            val nameLength = name.length
            val presentableTextLength = presentableText.length
            if (nameLength <= presentableTextLength) {
                lookupElementPresentation.appendTailText(presentableText.substring(nameLength), true)
            }
        }
        val locationString = itemPresentation.locationString
        if (locationString != null) {
            lookupElementPresentation.appendTailText(" ($locationString)", false)
        }
    }

    private fun renderObject(lookupElement: LookupElement) {
        val logger = Logger.getInstance(CallDefinitionClause::class.java)
        val `object` = lookupElement.getObject()
        val title = "CallDefinitionClause render called on LookupElement with null getPsiElement"
        val message = """
            ## name
            
            ```
            $name
            ```
            
            ## getObject()
            
            ### toString()
            
            ```
            $`object`
            ```
            
            ### getClass().getName()
            
            ```
            ${`object`.javaClass.name}
            ```
            
            """.trimIndent()
        logger.error(message, Throwable(title))
    }
}
